# Probability (05_Probability.ipynb)

**Sprint:** Sprint 2 – Statistics & Mathematics for AI/ML Engineers

## 1. Introduction to Probability

### Definition
Probability measures how LIKELY an event is to occur, expressed as 
a number between 0 (impossible) and 1 (certain). It's the 
mathematical foundation for reasoning about uncertainty.

### Formula
P(Event) = (Number of favorable outcomes) / (Total number of possible outcomes)

### Numerical Example
Rolling a fair die and wanting a 4: 
P(rolling a 4) = 1/6 ≈ 0.167

### Business Example
An insurance company calculates the probability of a claim being 
filed to price premiums appropriately - higher probability of 
claims means higher premiums needed to remain profitable.

### AI/ML Example
Every classification model (like a spam filter) outputs a 
PROBABILITY that an email is spam, not just a hard yes/no - 
probability is the foundation of nearly all predictive modeling.

### Interpretation
A probability of 0.167 for rolling a 4 means if you rolled the die 
many times, roughly 16.7% of rolls would show a 4 - probability 
describes long-run frequency, not a guarantee for any single roll.

In [1]:
# Simulating probability of rolling a 4, many times
import random

trials = 10000
count_4 = sum(1 for _ in range(trials) if random.randint(1, 6) == 4)
simulated_probability = count_4 / trials

print("Simulated probability of rolling a 4:", simulated_probability)
print("Theoretical probability:", 1/6)

# Where it's used: rolling a simulated die 10,000 times and
# counting how often we got a 4 gives an EMPIRICAL probability that
# should be close to the THEORETICAL probability (1/6 ≈ 0.167) -
# demonstrating probability as long-run frequency in action.

Simulated probability of rolling a 4: 0.1661
Theoretical probability: 0.16666666666666666


## 2. Events and Sample Space

### Definition
Sample Space is the set of ALL possible outcomes of an experiment. 
An Event is a specific outcome or set of outcomes we're interested 
in - a SUBSET of the sample space.

### Formula
Sample Space (S) = {all possible outcomes}
Event (E) = {specific outcome(s) of interest}, where E ⊆ S

### Numerical Example
Rolling a die: Sample Space = {1, 2, 3, 4, 5, 6}
Event "rolling an even number" = {2, 4, 6} - a subset of the sample space

### Business Example
For a product launch, the Sample Space might be all possible 
customer reactions (buy, ignore, complain, refer a friend). The 
Event "customer buys" is one specific outcome the business tracks.

### AI/ML Example
In classification problems, the Sample Space is all possible class 
labels a model could predict (e.g., {cat, dog, bird}), and a 
specific prediction is one Event from that space.

### Interpretation
Since Event = {2,4,6} has 3 outcomes out of a Sample Space of 6 
total outcomes, P(even number) = 3/6 = 0.5 - understanding sample 
space is necessary before we can even calculate probability correctly.

In [2]:
sample_space = [1, 2, 3, 4, 5, 6]
event_even = [x for x in sample_space if x % 2 == 0]

probability_even = len(event_even) / len(sample_space)
print("Sample Space:", sample_space)
print("Event (even numbers):", event_even)
print("Probability of even number:", probability_even)

# Where it's used: sample_space lists ALL 6 possible outcomes.
# event_even filters out just the outcomes we care about (2, 4, 6) -
# a genuine SUBSET. Dividing the event's size by the sample space's
# size gives the probability directly, tying back to the formula above.

Sample Space: [1, 2, 3, 4, 5, 6]
Event (even numbers): [2, 4, 6]
Probability of even number: 0.5


## 3. Independent Events and Dependent Events

### Definition
Independent Events are events where the outcome of ONE does NOT 
affect the probability of the other. Dependent Events are the 
opposite - the outcome of one event DOES affect the probability of 
the other happening.

### Formula
Independent: P(A and B) = P(A) × P(B)
Dependent: P(A and B) = P(A) × P(B|A)   [B given A already happened]

### Numerical Example
Independent: Flipping a coin twice. P(Heads then Heads) = 0.5 × 0.5 = 0.25
Dependent: Drawing 2 cards from a deck WITHOUT replacement. 
P(first Ace) = 4/52. P(second Ace | first was Ace) = 3/51 (fewer 
aces and cards left)

### Business Example
Independent: Whether it rains today and whether a customer buys a 
product online (unrelated events). 
Dependent: Whether a customer received a discount coupon and 
whether they make a purchase (the coupon likely INFLUENCES the purchase).

### AI/ML Example
Naive Bayes classifiers assume features are INDEPENDENT of each 
other (a simplifying assumption, hence "naive") - even though in 
reality, many features are actually dependent on each other.

### Interpretation
For independent events, knowing one outcome tells you NOTHING 
about the other. For dependent events, knowing one outcome changes 
your probability estimate for the other - this distinction is 
critical for correctly calculating combined probabilities.

In [3]:
# Independent: two coin flips
p_heads = 0.5
p_two_heads_independent = p_heads * p_heads
print("P(two heads in a row - independent):", p_two_heads_independent)

# Dependent: drawing 2 aces without replacement from a 52-card deck
p_first_ace = 4/52
p_second_ace_given_first = 3/51   # one ace and one card already removed
p_two_aces_dependent = p_first_ace * p_second_ace_given_first
print("P(two aces in a row - dependent):", p_two_aces_dependent)

# Where it's used: the independent case simply multiplies two FIXED
# probabilities (0.5 x 0.5), since one flip doesn't affect the
# other. The dependent case uses a CHANGED probability for the
# second draw (3/51, not 4/52), since removing the first ace
# genuinely altered what's left in the deck.

P(two heads in a row - independent): 0.25
P(two aces in a row - dependent): 0.004524886877828055


## 4. Conditional Probability

### Definition
Conditional Probability is the probability of an event occurring, 
GIVEN that another event has already happened. Written as P(A|B) - 
"probability of A given B."

### Formula
P(A|B) = P(A and B) / P(B)

### Numerical Example
In a class of 30 students, 18 like Math, 12 like Science, and 8 
like both. 
P(likes Science | likes Math) = P(both) / P(Math) = 8/30 / 18/30 = 8/18 ≈ 0.44

### Business Example
An e-commerce company calculates P(purchase | added to cart) - the 
probability a customer completes a purchase GIVEN they already 
added an item to their cart, which is very different from the 
overall purchase probability across all site visitors.

### AI/ML Example
Conditional Probability is the mathematical foundation of Naive 
Bayes classifiers, which calculate P(class | features) - the 
probability of a class label GIVEN the observed feature values.

### Interpretation
A conditional probability of 0.44 (liking Science given liking 
Math) tells us there's meaningful overlap between the two 
interests - conditional probability lets us update our estimate 
based on relevant additional information, rather than using a 
generic overall probability.

In [4]:
# Conditional probability example
total_students = 30
likes_math = 18
likes_science = 12
likes_both = 8

p_math = likes_math / total_students
p_both = likes_both / total_students

p_science_given_math = p_both / p_math
print("P(likes Science | likes Math):", round(p_science_given_math, 3))

# Where it's used: p_both (the overlap) divided by p_math (the
# 'given' condition) directly implements the conditional
# probability formula - notice this differs from the plain
# probability of liking Science overall (12/30 ≈ 0.4), since we're
# now conditioning on students who already like Math.

P(likes Science | likes Math): 0.444


## 5. Bayes' Theorem

### Definition
Bayes' Theorem lets you UPDATE a probability estimate when new 
evidence becomes available. It flips a conditional probability 
around - if you know P(B|A), it lets you calculate P(A|B), which 
is often what you actually want to know.

### Formula
P(A|B) = [P(B|A) × P(A)] / P(B)

Where:
- P(A) = prior probability (before seeing evidence)
- P(B|A) = likelihood (probability of evidence, given A is true)
- P(B) = total probability of the evidence
- P(A|B) = posterior probability (updated belief, after seeing evidence)

### Numerical Example
A disease affects 1% of people (P(Disease) = 0.01). A test correctly 
detects it 99% of the time (P(Positive|Disease) = 0.99), but also 
gives a false positive 5% of the time for healthy people 
(P(Positive|No Disease) = 0.05).

If someone tests POSITIVE, what's the actual probability they have the disease?

P(Positive) = P(Positive|Disease)×P(Disease) + P(Positive|No Disease)×P(No Disease)
            = (0.99×0.01) + (0.05×0.99) = 0.0099 + 0.0495 = 0.0594

P(Disease|Positive) = (0.99 × 0.01) / 0.0594 ≈ 0.1667 (only ~16.7%!)

### Business Example
A spam filter uses Bayes' Theorem: given that an email contains 
the word "free" (evidence), what's the probability it's spam? It 
combines how often "free" appears in spam vs legitimate emails 
with the overall spam rate, to update the probability.

### AI/ML Example
Naive Bayes classifiers are directly built on this formula - they 
calculate P(class | features) by combining the prior probability 
of each class with the likelihood of observing those specific 
feature values for that class.

### Interpretation
Even with a 99% accurate test, a positive result only means ~16.7% 
chance of actually having the disease - because the disease is RARE 
(1%), most positive results are actually false positives from the 
much larger healthy population. This is the counter-intuitive power 
of Bayes' Theorem - it corrects our intuition using the full picture, 
not just the test's accuracy alone.

In [5]:
# Bayes' Theorem: disease testing example
p_disease = 0.01
p_positive_given_disease = 0.99
p_positive_given_no_disease = 0.05
p_no_disease = 1 - p_disease

# Total probability of testing positive (any reason)
p_positive = (p_positive_given_disease * p_disease) + (p_positive_given_no_disease * p_no_disease)

# Bayes' Theorem - what we actually want to know
p_disease_given_positive = (p_positive_given_disease * p_disease) / p_positive

print("P(Positive test):", round(p_positive, 4))
print("P(Disease | Positive test):", round(p_disease_given_positive, 4))
print("In percentage:", round(p_disease_given_positive * 100, 2), "%")

# Where it's used: p_positive combines BOTH ways a positive test can
# happen (true positive from sick people + false positive from
# healthy people). Dividing the TRUE positive portion by this total
# gives the actual updated probability - notice it's much lower
# than the test's 99% accuracy might suggest, because the disease
# is so rare to begin with.

P(Positive test): 0.0594
P(Disease | Positive test): 0.1667
In percentage: 16.67 %


## Interview Questions - Probability

**Q1: What is the Sample Space of an experiment?**
The set of ALL possible outcomes of that experiment - e.g., rolling 
a die has a sample space of {1,2,3,4,5,6}.

**Q2: What's the difference between an Event and Sample Space?**
Sample Space is every possible outcome; an Event is a specific 
subset of outcomes we're interested in (e.g., "rolling an even 
number" is an event within the full sample space).

**Q3: What's the difference between Independent and Dependent Events?**
Independent events don't affect each other's probability (e.g., 
two separate coin flips). Dependent events DO affect each other 
(e.g., drawing cards without replacement).

**Q4: How do you calculate the probability of two independent 
events both happening?**
Multiply their individual probabilities: P(A and B) = P(A) × P(B).

**Q5: What is Conditional Probability?**
The probability of an event happening GIVEN that another event has 
already occurred, written as P(A|B) = P(A and B) / P(B).

**Q6: State Bayes' Theorem and explain what each term means.**
P(A|B) = [P(B|A) × P(A)] / P(B). P(A) is the prior belief, P(B|A) 
is the likelihood, P(B) is the total probability of the evidence, 
and P(A|B) is the updated (posterior) probability.

**Q7: Why can a 99%-accurate medical test still give mostly false 
positives for a rare disease?**
Because the disease is rare, the large number of healthy people 
tested produces many false positives, which can outnumber the true 
positives from the small group of actually sick people - Bayes' 
Theorem reveals this.

**Q8: What is a Random Variable?**
A variable whose value is the numerical outcome of a random 
process or experiment (e.g., the number shown when rolling a die).

**Q9: What is Expected Value?**
The long-run average outcome of a random variable if an experiment 
were repeated many times, calculated as the sum of each outcome 
multiplied by its probability.

**Q10: Why is Bayes' Theorem important in Machine Learning?**
It's the mathematical foundation of Naive Bayes classifiers, which 
calculate the probability of a class label given observed feature 
values (P(class | features)) - directly applying Bayes' formula.